### DISCUSSION TITLE
Rider works fine but VS Code I can only run Android and no choice for iOS

### DISCUSSION THREAD

**Discussion by stephenhauck** [unoplatform/uno#15563] [Q&A]
I'm running on a Mac ...

I open the same project with Rider and VS Code and it will let me run iOS or Android in Rider but not in VS Code ... it only offers the Android option .. what am I missing ?

I would like the option to use either tool after the project is created 

Thanks in advance for ANY help.

**Comment by spouliot** [comment]
In the status bar you can click on the TFM selector (e.g. `net8.0-android | Debug`) to see the list of targets available in your project.

<img width="1000" alt="Screenshot 2024-02-20 at 5 28 41 PM" src="https://github.com/unoplatform/uno/assets/260465/b82520e9-0492-46df-9e36-8dce668d1c79">

Select `netX.0-ios | Debug` target then click on the next status bar item to choose a simulator or a connected device.

When you start debugging the project will be compiled for the selected OS/device then launched.

If that does not work then have a look at the `Uno Platform - Debugger` logs. This is located inside the **Output** and there's a combobox to select the `Uno Platform - Debugger` logs.

<img width="1397" alt="Screenshot 2024-02-20 at 5 33 47 PM" src="https://github.com/unoplatform/uno/assets/260465/2d52ae0d-ad97-4b01-bfb8-8fa82109fd44">

Please copy/paste (as text) the contents of the logs so we can see what went wrong.

**Reply by stephenhauck** [reply]
this is all I see 
<img width="674" alt="image" src="https://github.com/unoplatform/uno/assets/12648793/712965ad-0cfd-4fce-8cce-723f3fbe34e6">

**Reply by stephenhauck** [reply]
<img width="847" alt="image" src="https://github.com/unoplatform/uno/assets/12648793/d42bf1db-b9dd-4b9d-a2c8-6a399e37f30f">

**Reply by stephenhauck** [reply]
<img width="940" alt="image" src="https://github.com/unoplatform/uno/assets/12648793/296a1664-0708-4465-b6e3-b25490c59e5f">

**Reply by spouliot** [reply]
That suggests the project was created solely for Android (no iOS or Mac[Catalyst]). However I don't see how Rider could make this work for iOS (since the platform specific code would be missing).

Can you paste the content of your `*.Mobile.csproj` file ?

**Reply by stephenhauck** [reply]
```xml
 <PropertyGroup>
    <TargetFrameworks>$(DotNetVersion)-android;$(DotNetVersion)-ios;$(DotNetVersion)-maccatalyst</TargetFrameworks>
    <TargetFrameworks Condition="'$(OverrideTargetFramework)'!=''">$(OverrideTargetFramework)</TargetFrameworks>
    <SingleProject>true</SingleProject>
    <OutputType>Exe</OutputType>
    <!-- Display name -->
    <ApplicationTitle>Uno_MVVM_ToDO_LiteDB</ApplicationTitle>
    <!-- App Identifier -->
    <ApplicationId>com.companyname.Uno</ApplicationId>
    <ApplicationIdGuid>F5FE7090-31EB-4595-A854-C654E473866C</ApplicationIdGuid>
    <!-- Versions -->
    <ApplicationDisplayVersion>1.0</ApplicationDisplayVersion>
    <ApplicationVersion>1</ApplicationVersion>

    <!-- Debugger workaround https://github.com/dotnet/maui-samples/blob/8aa6b8780b12e97b157514c3bdc54bb4a13001cd/HelloMacCatalyst/HelloMacCatalyst.csproj#L7 -->
    <!-- <MtouchExtraArgs Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'maccatalyst'">$(MtouchExtraArgs) -setenv:MONO_THREADS_SUSPEND=preemptive</MtouchExtraArgs> -->
    <!-- Required for C# Hot Reload -->
    <UseInterpreter Condition="'$(Configuration)' == 'Debug' and $([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) != 'maccatalyst'">True</UseInterpreter>
    <IsUnoHead>true</IsUnoHead>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
    <CodesignKey>Apple Development: Stephen Hauck (JRWV3ZBCGQ)</CodesignKey>
  </PropertyGroup>
```

**Reply by stephenhauck** [reply]
<img width="516" alt="image" src="https://github.com/unoplatform/uno/assets/12648793/600bfd76-41fb-41de-a083-75ede7441d8a">

**Reply by spouliot** [reply]
The project file seems fine. 

1. The list of TFM comes from (Microsoft) C# Dev Kit extension. Can you show me (screenshot) what the **Solution Explorer** (in the left side bar) looks like ?
Mine looks like:

<img width="602" alt="Screenshot 2024-02-20 at 7 18 29 PM" src="https://github.com/unoplatform/uno/assets/260465/8b2213a6-1137-4bc4-b81c-27c3a64d6209">

So we can see Android, iOS, MacCatalyst under the `Source / Platforms`

2. Please run `uno-check` from the command-line and paste the output (as text) in a reply.
It's possible the iOS workload is not installed (or incompatible with the current dotnet 8.0 you're using).

3. Please copy/paste (as text) the contents of the `Uno Platform` logs so we can see what went wrong (just like the last screenshot from my first reply, but do it for `Uno Platform`, not `Uno Platform - Debugger`).

thanks!

**Reply by stephenhauck** [reply]
[Info] Launching XAML language server /Users/stephenhauck/.vscode/extensions/unoplatform.vscode-0.12.0/bin/lshost/net8.0/Uno.LSP.Host.dll
[Info] Found 'ms-dotnettools.csharp' version 2.15.30.
[Info] Found 'ms-dotnettools.csdevkit' version 1.3.10.
[Info] Found 'ms-dotnettools.dotnet-maui' version 0.8.44.
[Info] C# extension onServerStateChange API is available.
[Info] Running in Dev Kit mode
[Info] Uno Platform "dotnet new" templates are already installed. The templates may be updated by running "dotnet new install Uno.Templates"
[Info] No project found inside /Users/stephenhauck/githubrepos/Uno_MVVM_ToDO_LiteDB/Uno_MVVM_ToDO_LiteDB.sln
[Info] Setting active project /Users/stephenhauck/githubrepos/Uno_MVVM_ToDO_LiteDB/Uno_MVVM_ToDO_LiteDB.Mobile/Uno_MVVM_ToDO_LiteDB.Mobile.csproj

**Reply by stephenhauck** [reply]
```
Pack Microsoft.NETCore.App.Runtime.Mono.maccatalyst-arm64 version 7.0.16 is already installed.
Writing workload pack installation record for Microsoft.NETCore.App.Runtime.Mono.net7.maccatalyst-arm64 version 7.0.16...
Installing pack Microsoft.NETCore.App.Runtime.Mono.maccatalyst-x64 version 7.0.16...
Pack Microsoft.NETCore.App.Runtime.Mono.maccatalyst-x64 version 7.0.16 is already installed.
Writing workload pack installation record for Microsoft.NETCore.App.Runtime.Mono.net7.maccatalyst-x64 version 7.0.16...
----------------
==workloadRollbackDefinitionJsonOutputStart==
{
  "microsoft.net.sdk.android": "34.0.79/8.0.100",
  "microsoft.net.sdk.ios": "17.2.8022/8.0.100",
  "microsoft.net.sdk.maccatalyst": "17.2.8022/8.0.100",
  "microsoft.net.sdk.macos": "14.2.8022/8.0.100",
  "microsoft.net.sdk.maui": "8.0.6/8.0.100",
  "microsoft.net.sdk.tvos": "17.2.8004/8.0.100",
  "microsoft.net.workload.mono.toolchain.current": "8.0.2/8.0.100",
  "microsoft.net.workload.emscripten.current": "8.0.2/8.0.100",
  "microsoft.net.workload.emscripten.net6": "8.0.2/8.0.100",
  "microsoft.net.workload.emscripten.net7": "8.0.2/8.0.100",
  "microsoft.net.workload.mono.toolchain.net6": "8.0.2/8.0.100",
  "microsoft.net.workload.mono.toolchain.net7": "8.0.2/8.0.100",
  "microsoft.net.sdk.aspire": "8.0.0-preview.2.23619.3/8.0.100"
}
==workloadRollbackDefinitionJsonOutputEnd==
  ✔ microsoft.net.sdk.android (34.0.79/8.0.100) installed.
  ✔ microsoft.net.sdk.ios (17.2.8022/8.0.100) installed.
  ✔ microsoft.net.sdk.maccatalyst (17.2.8022/8.0.100) installed.
  ✔ microsoft.net.sdk.macos (14.2.8022/8.0.100) installed.
  ✔ microsoft.net.sdk.maui (8.0.6/8.0.100) installed.
────────────────────────────────────────────────────────────────────────────────────────────────────────────

✔ Congratulations, everything looks great!

Press Enter to finish...
```

**Reply by stephenhauck** [reply]
<img width="511" alt="image" src="https://github.com/unoplatform/uno/assets/12648793/f9b7f53e-e6e7-47df-84de-982d1ad7db60">

**Reply by stephenhauck** [reply]
<img width="678" alt="image" src="https://github.com/unoplatform/uno/assets/12648793/c11c2f3e-4161-4b3d-a1a5-9ba718bb2290">

**Reply by stephenhauck** [reply]
<img width="516" alt="image" src="https://github.com/unoplatform/uno/assets/12648793/5ce4c5f8-0b9b-4889-a8b7-3f5929f3f79c">

**Reply by spouliot** [reply]
~~Can you try to switch to the pre-release version of C# DevKit ?
Once updated, reload the window and see if you get all the TFM choices.~~

Try deleting the `*.csproj.user` file and reload the (VSCode) window (press `F1` and select `Developer: Reload Window`)

**Reply by stephenhauck** [reply]
that helped but it seems it's trying to run on X64 and not ARM 64 as I have selected  .. I am running on an M1 Max


 *  Executing task: dotnet build "/Users/stephenhauck/githubrepos/Uno_MVVM_ToDO_LiteDB/Uno_MVVM_ToDO_LiteDB.Mobile/Uno_MVVM_ToDO_LiteDB.Mobile.csproj" -f:net8.0-ios -c:Debug -r:iossimulator-x64 -clp:NoSummary -p:GenerateFullPaths=true -p:UnoForceSingleTFM=true -bl:"/Users/stephenhauck/githubrepos/Uno_MVVM_ToDO_LiteDB/Uno_MVVM_ToDO_LiteDB.Mobile/net8.0-ios-Debug-iossimulator-x64.binlog" 
 *

**Reply by stephenhauck** [reply]
<img width="675" alt="image" src="https://github.com/unoplatform/uno/assets/12648793/e98c9eed-218a-4798-a99b-7944cadb8ae9">

**Reply by stephenhauck** [reply]
Now I relaunched and it didn't find the Android SDK or XCode

**Reply by stephenhauck** [reply]
I deleted the file again and restarted and it seems to see everything again but it still is trying to build for X64 and not ARM64

**Reply by stephenhauck** [reply]
/Users/stephenhauck/githubrepos/Uno_MVVM_ToDO_LiteDB/Uno_MVVM_ToDO_LiteDB.Mobile/bin/Debug/
bin/Debug/net8.0-ios/iossimulator-x64/Uno_MVVM_ToDO_LiteDB.Mobile.app could not be installed on simulator 0D0B6343-63FC-48DE-A07B-FBDC6EC8AFF6.
An error was encountered processing the command (domain=NSPOSIXErrorDomain, code=2):
Simulator device returned an error for the requested operation.
An application bundle was not found at the provided path.
Provide a valid path to the desired application bundle.
Underlying error (domain=NSPOSIXErrorDomain, code=2):
	Failed to install the requested application
	An application bundle was not found at the provided path.

**Reply by stephenhauck** [reply]
<img width="278" alt="image" src="https://github.com/unoplatform/uno/assets/12648793/7f16b5e1-0dcf-47d1-86eb-7855c8da912b">

**Reply by stephenhauck** [reply]
I have to call it a day .. will check back first thing in the morning ...

**Reply by spouliot** [reply]
ok, I found out what the problem is. Until a fix is released you'll need to delete that file `*.user` file when you restart VScode (at least if you need to change the target).

> that helped but it seems it's trying to run on X64 and not ARM 64 as I have selected .. I am running on an M1 Max

That's a mix of two different issues.

First it's a bit confusing (from Apple) since the simulator process can run both `arm64` and `x64` applications (at least on `arm64` computers using Rosetta).

So even if you ask for an `arm64` simulator it will be able to run `x64` applications. IMO it's generally useful but it can also be surprising.


The second part is a known issue.

SkiaSharp 2.88.3 (and older) ships without binaries for `iossimulator-arm64` and Uno applications requires SkiaSharp. That requirement is _indirect_ (so you don't see it in your application's `.csproj` file). This means SkiaSharp is transitively restored based on other assemblies. Sadly (for now) the incomplete 2.88.3 is being picked up and projects wouldn't build properly.

Once this is resolved the VSCode extension will be updated to honour your selection of `arm64` or `x64` simulators.

**Reply by jeromelaban** [reply]
@stephenhauck thanks for the troubleshooting! About SkiaSharp, we have an update coming soon in the templates, but in the meantime, you can:

- Install both packages in "SkiaSharp.Skottie" and "SkiaSharp.Views.Uno.WinUI" on the common project
- Install "SkiaSharp.NativeAssets.iOS" and "SkiaSharp.NativeAssets.Android" in the `.Mobile` project

You can use the latest version of SkiaSharp for all these packages.

**Reply by sasakrsmanovic** [reply]
Thanks for all the troubleshooting @stephenhauck !!

**Reply by spouliot** [reply]
@stephenhauck there's a new (0.13) version of the VSCode extension available. It solve the problem with the `*.csproj.user` file (so you don't have to delete it to select iOS) and workaround another [issue specific to dotnet 8.0.200](https://github.com/dotnet/sdk/issues/38932).

The [fix for a newer Skia](https://github.com/unoplatform/uno/pull/15567) is in progress. On landed a a future version of the VS Code extension will be able to allow arm64 process for the iOS simulator.
